### PR TITLE
feat(codec): implement ParamValue payload v1 and batch codec (#4)

### LINKED ISSUE #4: [M1] ParamValue and payload v1 codec
## Summary

Implement `ParamValue` (the typed value class) and the payload v1
encoder/decoder, verified byte-for-byte against the golden fixtures.

## References

- `docs/04_api_cpp.md` §1 (`ParamValue` API, arithmetic cast rules)
- `docs/03_wire_protocol.md` §2 (payload v1 layout), §2.3 (golden fixtures),
  §5 (batch format, §5.1 batch fixture)

## Scope

- [ ] `include/sitos/param_value.hpp` + `src/param_value.cpp`:
      `ParamValue` (variant of bool/int64/double/string/bytes),
      `type()`, `As<T>()` with Bool/S64/Dp arithmetic casts,
      `AsSpan<T>()` (zero-copy bytes view), `Encode()`, `Decode()`
- [ ] Batch codec: encode/decode of the `sitos.v1.batch` layout
      (`include/sitos/batch.hpp`, `src/batch.cpp` — codec only, no transport)
- [ ] Golden fixtures as files: `tests/fixtures/payload_v1/*.hex`
      matching `docs/03` §2.3 and §5.1 exactly
- [ ] Independent fixture validation: a Python script using only `struct`
      regenerates the fixtures and asserts byte equality (guards against
      a wrong fixture being baked into both impl and test)
- [ ] `tests/unit/param_value_test.cpp`: `PayloadV1GoldenFixtures`,
      `BatchV1GoldenFixture`, full round-trips, NaN/±inf, empty string/bytes,
      truncated/invalid payload rejection, canonical NaN normalization

## Out of scope

- Key handling (#5), any zenoh usage

## Acceptance Criteria

1. `PayloadV1GoldenFixtures` and `BatchV1GoldenFixture` are green
2. Round-trip tests cover all five types including boundary values
3. Invalid payloads (empty, unknown tag, truncated) are rejected without UB
   (verified under ASan/UBSan)
4. The Python fixture validator passes in CI

## Dependencies

- #1

## For the implementer

- Copy the checklist above into your PR description and tick items as you
  complete them. Do not edit the checklist in this issue.
- Follow `docs/development_workflow.md` (branch `issue/4-param-value-codec`,
  Conventional Commits per §2.1, TDD: fixtures tests RED first).
- This issue is closed automatically by the PR via `Closes #4`.

### PR DESCRIPTION
**Pull request by tetsuh**
## Summary

Closes #4

## Requirements

- C01: wire-compatible payload v1 (golden fixtures)
- F09: batch atomic delivery (codec only, no transport)
- X01: contract-test principle (golden fixture parity C++ ↔ Python)
- C04: versioning via `sitos.v1` / `sitos.v1.batch`

## Checklist (from #4)

- [x] `include/sitos/param_value.hpp` + `src/param_value.cpp`: `ParamValue`
  variant, `type()`, `As<T>()` with Bool/S64/Dp casts, `AsSpan<T>()`,
  `Encode()`, `Decode()`
- [x] Batch codec: `include/sitos/batch.hpp` + `src/batch.cpp` (codec only,
  no transport)
- [x] Golden fixtures `tests/fixtures/payload_v1/*.hex` matching `docs/03`
  §2.3 and §5.1 exactly
- [x] Independent fixture validation via Python `struct`
  (`tests/fixtures/validate_fixtures.py`)
- [x] `tests/unit/param_value_test.cpp`: `PayloadV1GoldenFixtures`,
  `BatchV1GoldenFixture`, round-trips, NaN/±inf, empty string/bytes,
  truncated/invalid rejection, canonical NaN normalization

### AC1 — golden fixtures green

Linux (gcc, CMake `dev-linux`, WSL2 Ubuntu 24.04.1 LTS):

```text
100% tests passed, 0 tests failed out of 9
Total Test time (real) =  0.65 sec
```

Windows (MSVC 19.44 + Ninja, `dev-windows`):

```text
100% tests passed, 0 tests failed out of 9
Total Test time (real) =  0.39 sec
```

### AC2 — round-trips cover all five types incl. boundaries

`ParamValue.RoundTripsAllTypesAndBoundaries` covers:
- BOOL (true / false)
- S64 ({0, -1, INT32_MAX, INT64_MIN, INT64_MAX})
- DP ({0, 240, 1e-300, 1e300, DBL_MAX, +inf, -inf, NaN})
- STR (empty / "abc" / "穀" / with control + NUL)
- BYTES (empty and [0x01, 0x02, 0xff])
plus re-encode byte stability after decode.

### AC3 — invalid payloads rejected without UB (verified under ASan/UBSan)

clang-18 + `-fsanitize=address,undefined`:

```text
[ RUN      ] ParamValue.RejectsInvalidPayloads          [ OK ]
[ RUN      ] ParamValue.RejectsMalformedBatches         [ OK ]
... 8 tests PASSED (ASan/UBSan clean)
```

Rejected cases: empty payload, reserved (5..7) and unused (0xFF) tags,
truncated/oversized S64 and DP bodies, BOOL body != 1 byte, batch count
truncated, key/value truncation, trailing bytes, unknown value tag.
Note: STR/BYTES with zero-length body remain VALID (str_empty, bytes_empty).

### AC4 — Python fixture validator passes in CI

```text
$ python tests/fixtures/validate_fixtures.py
OK   bool_false ... OK   bytes_0102ff   OK   batch_base_two_entries
PASS (fixture validator): 13 singles + 1 batch
```

Added as a CI step in both `build-linux` and `build-windows` jobs
(`.github/workflows/ci.yml`). It rebuilds every fixture from first
principles using only `struct` + UTF-8 and asserts byte equality with the
`.hex` files, so a wrong fixture baked into both the C++ impl and the C++
test would still be caught.

## RED-phase Confirmation

The codec tests were written first and confirmed RED against an unimplemented
`Encode()` stub:

```text
The following tests FAILED:
      2 - PayloadV1GoldenFixtures.SingleValuesMatchSpecByteForByte (Failed)
      3 - PayloadV1GoldenFixtures.NanIsCanonicalOnReencode (Failed)
      4 - BatchV1GoldenFixture.EncodesAndDecodesTheSpecFixture (Failed)
      5 - ParamValue.RoundTripsAllTypesAndBoundaries (Failed)
44% tests passed, 5 tests failed out of 9
```

This confirmed the tests actually exercise the codec. After implementing the
real `Encode()` / `Decode()` / batch encoder, the stub was removed and all 9
tests went GREEN.

During RED I caught one genuine implementation bug and one test bug, both
fixed before GREEN:

1. **Bug (impl):** `EncodeBatch` originally omitted the `vLen` field
   (the `sitos.v1.batch` layout in `docs/03` §5 is
   `[kLen][key][tag][vLen][body]`). Symptom: the encoded batch was 48 B vs the
   expected 56 B (missing two 4 B `vLen` fields). Fixed.
2. **Bug (test):** the AsSpan test compared the span pointer against the
   caller's input vector. Zero-copy actually means a view over `ParamValue`'s
   own internal buffer (a separate allocation, since the constructor copies).
   Fixed the assertion to check stable internal storage across repeated calls.

## ADR Needed?

- [x] No. This implements the wire format already specified in
  `docs/03_wire_protocol.md` under existing **ADR-0007**. No fixture meanings
  or new wire decisions were introduced.

### CODE REVIEW

**Comment by tetsuh** [comment]
@coderabbitai review

**Comment by tetsuh** [comment]
@coderabbitai review

File: CMakeLists.txt [hunk 1]
```diff
@@ -15,18 +15,22 @@ option(SITOS_BUILD_EXAMPLES "Build examples" OFF)
 
 # Warnings as errors
 if(MSVC)
-  add_compile_options(/W4 /WX)
+  add_compile_options(/W4 /WX /utf-8)
 else()
   add_compile_options(-Wall -Wextra -Wpedantic -Werror)
 endif()
```
[the remark this answers is not in the document]
Comment (tetsuh) [reply CMakeLists.txt]: **Fixed in 0b89a78** — moved warning flags from `add_compile_options` to a per-target `sitos_enable_warnings(target)` helper using `target_compile_options(... PRIVATE …)`, applied to `sitos` and both test targets. Vendored GoogleTest is no longer pinned to `-Werror`.

File: CMakeLists.txt [hunk 2]
```diff
@@ -15,18 +15,22 @@ option(SITOS_BUILD_EXAMPLES "Build examples" OFF)
...
+  src/param_value.cpp
+  src/batch.cpp
+)
 add_library(sitos::sitos ALIAS sitos)
-target_include_directories(sitos INTERFACE
+target_include_directories(sitos PUBLIC
   $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
   $<INSTALL_INTERFACE:include>
 )
+target_compile_features(sitos PUBLIC cxx_std_20)
```
[the remark this answers is not in the document]
Comment (tetsuh) [reply CMakeLists.txt:38]: **Skipped (out of scope for #4).** This PR's scope is the payload v1 / batch codec and its tests; ACs are golden fixtures, round-trip, invalid-reject without UB, and the Python validator. install/export rules and `find_package(sitos)` consumer wiring belong to the packaging issue and will land in a separate PR. CC @tetsuh.

File: include/sitos/param_value.hpp [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+  std::optional<std::span<const T>> AsSpan() const {
+    static_assert(std::is_trivially_copyable_v<T>,
+                  "AsSpan requires a trivially copyable element type");
+    auto* p = std::get_if<std::vector<std::byte>>(&value_);
+    if (p == nullptr) return std::nullopt;
+    if (p->size() % sizeof(T) != 0) return std::nullopt;
+    if (p->empty()) return std::span<const T>{};
+    auto* data = reinterpret_cast<const T*>(p->data());
+    return std::span<const T>{data, p->size() / sizeof(T)};
+  }
```
[the remark this answers is not in the document]
Comment (tetsuh) [reply include/sitos/param_value.hpp:117]: **Fixed in 0b89a78** — `AsSpan` now also checks `reinterpret_cast<uintptr_t>(p->data()) % alignof(T) != 0` and returns `std::nullopt` on mis-aligned storage, preventing UB for over-aligned `T`.

File: src/batch.cpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    off += *k_len;
+
+    if (off >= payload.size()) return std::nullopt;
+    std::uint8_t tag = static_cast<std::uint8_t>(payload[off]);
+    off += 1;
+
+    auto v_len = ReadLeU32(payload, off);
+    if (!v_len) return std::nullopt;
+    if (off + *v_len > payload.size()) return std::nullopt;
+
```
[the remark this answers is not in the document]
Comment (tetsuh) [reply src/batch.cpp:77]: **Fixed in 0b89a78** — both length guards now use the remaining-capacity form `*k_len > payload.size() - off` (and `*v_len > payload.size() - off`), avoiding unsigned wraparound on small `size_t`. Added the missing `<algorithm>` for `std::min` as well.

File: tests/fixtures/validate_fixtures.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    batch_expected = expected_batch()
+    if batch_actual != batch_expected:
+        failures += 1
+        print(f"FAIL batch_base_two_entries: expected {batch_expected.hex(' ')}\n"
+              f"     got {batch_actual.hex(' ')}")
+    else:
+        print("OK   batch_base_two_entries")
+
+    print(f"PASS (fixture validator): {len(singles)} singles + 1 batch")
+    return 1 if failures else 0
```
[the remark this answers is not in the document]
Comment (tetsuh) [reply tests/fixtures/validate_fixtures.py:100]: **Fixed in 0b89a78** — the validator now prints `PASS`/`FAIL` based on the failure count (e.g. `PASS (fixture validator): 13 singles + 1 batch, 0 failure(s)`).

File: tests/unit/param_value_test.cpp [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+    auto reencoded = decoded->Encode();
+    bool stable = BytesEq(encoded, reencoded);
+    if (!stable) decoded = std::nullopt;
+    return decoded;
+  };
+
+  // BOOL
+  ASSERT_TRUE(roundtrip(sitos::ParamValue(false))->As<bool>().has_value());
+  EXPECT_FALSE(*roundtrip(sitos::ParamValue(false))->As<bool>());
+  EXPECT_TRUE(*roundtrip(sitos::ParamValue(true))->As<bool>());
```
[the remark this answers is not in the document]
Comment (tetsuh) [reply tests/unit/param_value_test.cpp]: **Fixed in 0b89a78** — both bool round-trip cases now do `ASSERT_TRUE(r.has_value())` and `ASSERT_TRUE(r->As<bool>().has_value())` before dereferencing, consistent with the S64/DP/STR/BYTES cases.

File: tests/unit/param_value_test.cpp [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+  }
+
+  // STR
+  for (const std::string& s : {std::string(""), std::string("abc"), std::string("穀"),
+                               std::string("multi\nline\x00with\tnul")}) {
+    auto r = roundtrip(sitos::ParamValue(s));
+    ASSERT_TRUE(r.has_value());
+    EXPECT_EQ(r->type(), sitos::ValueType::Str);
+    EXPECT_EQ(*r->As<std::string>(), s);
+  }
```
[the remark this answers is not in the document]
Comment (tetsuh) [reply tests/unit/param_value_test.cpp:200]: Good catch — real test bug. **Fixed in 0b89a78** — built the NUL string as `std::string("multi\nline\x00with\tnul", 19)` so the `const char*` ctor no longer truncates at the embedded `\x00`; the round-trip now actually exercises the embedded NUL.

File: tests/unit/param_value_test.cpp [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+  expect_invalid({std::byte{0x01}, std::byte{0x00}, std::byte{0x00}, std::byte{0x00},
+                  std::byte{0x00}, std::byte{0x00}, std::byte{0x00}, std::byte{0x04},
+                  std::byte{0x00}, std::byte{0x00}, std::byte{0x00}, std::byte{0x01}});
+  // unknown value tag in an entry
+  std::vector<std::byte> bad{
+      std::byte{0x01}, std::byte{0}, std::byte{0}, std::byte{0}, std::byte{0},
+      // kLen=0, tag=0xFE (unknown), vLen=0
+      std::byte{0}, std::byte{0xFE}, std::byte{0}, std::byte{0}, std::byte{0}, std::byte{0}};
+  expect_invalid(bad);
+}
```
[the remark this answers is not in the document]
Comment (tetsuh) [reply tests/unit/param_value_test.cpp:322]: Good catch — real test bug. **Fixed in 0b89a78** — the trailing case is now one fully valid entry (count=1, kLen=0, BOOL, vLen=1, body=0x00) + an extra 0xFF; the unknown-tag case is kLen=0, tag=0xFE, vLen=0. Each now reaches its intended `DecodeBatch` failure path (trailing-bytes / `ParamValue::Decode` unknown-tag) instead of the key-length guard.

**Comment by tetsuh** [comment]
## CodeRabbit review — all addressed

Push: `0b89a78` (CI build-linux ✅ build-windows ✅).

**Actionable comments (8)**
- CMake warnings global → target-scoped: **fixed**
- CMake install/export rules: **skipped — out of #4 scope** (packaging issue, separate PR)
- `AsSpan` alignment UB: **fixed** (`alignof(T)` check)
- `batch.cpp` unsigned wraparound in length guards: **fixed** (remaining-capacity form)
- `validate_fixtures.py` prints PASS on failure: **fixed**
- bool round-trip `has_value()`: **fixed**
- embedded-NUL test silently truncated by `const char*` ctor: **fixed** (real test bug) — length-prefixed string
- malformed batch trailing / unknown-tag reached key-length guard first: **fixed** (real test bug) — valid entry first

**Nitpicks (2) — also applied**
- `[[nodiscard]]` on `EncodeBatch` / `DecodeBatch`: **applied**
- missing `<algorithm>` for `std::min` in `src/batch.cpp`: **applied**

**Verification:** Linux gcc ✅ (9/9), clang-18 + ASan/UBSan ✅ (8/8, clean), Windows MSVC 19.44 + Ninja ✅ (9/9), Python validator ✅ (`PASS … 0 failure(s)`).

Two of the findings were genuine test bugs that would have shipped false-green coverage (the embedded-NUL test never exercised an embedded NUL; the trailing/unknown-tag rejection tests never reached the rejection branch they claimed). Good catch, CodeRabbit.
